File: csharp/patterns/src/PrototypePatterns/Structural/Flyweight/FlyweightPattern.cs
namespace PrototypePatterns.Structural.Flyweight;

public sealed record GlyphStyle(string Family, int Size, bool Bold);

public sealed class GlyphStyleFactory
{
    private readonly Dictionary<string, GlyphStyle> _cache = [];

    public GlyphStyle GetStyle(string family, int size, bool bold)
    {
        var key = $"{family}:{size}:{bold}";
        if (!_cache.TryGetValue(key, out var style))
        {
            style = new GlyphStyle(family, size, bold);
            _cache[key] = style;
        }

        return style;
    }
}

public static class FlyweightPattern
{
    public static bool Demo()
    {
        var factory = new GlyphStyleFactory();
        var first = factory.GetStyle("Fira Code", 12, false);
        var second = factory.GetStyle("Fira Code", 12, false);
        return ReferenceEquals(first, second);
    }
}

File: java/patterns/src/main/java/org/prototypepatterns/structural/flyweight/FlyweightPattern.java
package org.prototypepatterns.structural.flyweight;

import java.util.HashMap;
import java.util.Map;

public final class FlyweightPattern {
    private FlyweightPattern() {
    }

    public record GlyphStyle(String family, int size, boolean bold) {
    }

    public static final class GlyphStyleFactory {
        private final Map<String, GlyphStyle> cache = new HashMap<>();

        public GlyphStyle getStyle(String family, int size, boolean bold) {
            String key = family + ":" + size + ":" + bold;
            return cache.computeIfAbsent(key, ignored -> new GlyphStyle(family, size, bold));
        }
    }

    public static boolean demo() {
        GlyphStyleFactory factory = new GlyphStyleFactory();
        GlyphStyle first = factory.getStyle("Fira Code", 12, false);
        GlyphStyle second = factory.getStyle("Fira Code", 12, false);
        return first == second;
    }
}
